### DISCUSSION TITLE
MultipartParseError: Did not find boundary character 48 at index 2 when uploading large files

### DISCUSSION THREAD

**Discussion by prashikdewtale10** [dymmond/esmerald#452] [Q&A] [answered]
I'm encountering the following error while attempting to upload files using multipart/form-data requests in the Esmerald framework 
`multipart.exceptions.MultipartParseError: Did not find boundary character 48 at index 2
`
This issue occurs when trying to upload a file, especially when the file size exceeds 1 MB or when the boundary is not correctly identified in the request.

I would greatly appreciate any assistance or suggestions on how to resolve this issue with file uploads exceeding 1 MB in the Esmerald framework. Thank you in advance for your time and help.

**Comment by tarsil** [comment]
Hi @prashikdewtale10 that is a new one. In order to be able to assist you, could you please provide more details in how you did implement it, like code examples please?

**Comment by tarsil** [comment]
@prashikdewtale10 I literally did this:

```python
from esmerald import Esmerald, post, Body, Gatewaym UploadFile
from esmerald.enums import EncodingType

@post("/upload-file")
async def upload(data: UploadFile = Body(media_type=EncodingType.MULTI_PART)) -> dict[str, str]:
    return {"name": data.filename}


app = Esmerald(
    routes=[
        Gateway(handler=upload),
    ]
)
```

You can also do:

```python
from esmerald import Esmerald, post, File, Gatewaym UploadFile

@post("/upload-file")
async def upload(data: UploadFile = File() -> dict[str, str]:
    return {"name": data.filename}


app = Esmerald(
    routes=[
        Gateway(handler=upload),
    ]
)
```

Tested even using the OpenAPI documentation at [https://localhost:8000/docs/swagger](https://localhost:8000/docs/swagger)

As per [documentation](https://www.esmerald.dev/extras/upload-files/?h=upload#single-file-upload) and uploaded a 7MB file and it works.

I'm not sure about your side of the implementation?

Apologies for taking so long but we have been quite busy.

**Comment by prashikdewtale10** [comment]
Hi,

I'm trying to implement a file upload API using the Esmerald framework. The API should accept a file and some contextual data (file_context and applicable_date) via form data submitted through Postman. Below is the current implementation:
```python
async def upload_files(
    request: Request,
    data: Any = Form(),
) -> Response:
    # Parse form data
    request_file = data.get('file')
    # for validating type of file_context it should be str only
    context_name = data.get("file_context")
    applicable_date = data.get("applicable_date")

app = Esmerald(
    routes=[
        Gateway(handler=upload_files),
    ]
)
```
What I'm Trying to Achieve

- Accept a file and additional form fields (file_context and applicable_date) via multipart form data.
- Validate that file_context is a string and handle any invalid inputs gracefully.
- Ensure efficient handling of file uploads, including error handling and response generation.

Issues and Questions

1. Is there a better way to handle and validate multipart form data in Esmerald?
2. How can I ensure that both the file and form data are processed securely and efficiently?
3. Are there any improvements or best practices I should adopt for this implementation?

**Reply by prashikdewtale10** [reply]
Traceback (most recent call last):
  File "d:\esmerald_venv\Lib\site-packages\esmerald\middleware\_exception_handlers.py", line 37, in wrapped_app
    await app(scope, receive, sender)
  File "d:\esmerald_venv\Lib\site-packages\esmerald\middleware\asyncexitstack.py", line 46, in __call__
    raise exception
  File "d:\esmerald_venv\Lib\site-packages\esmerald\middleware\asyncexitstack.py", line 38, in __call__
    await self.app(scope, receive, send)
  File "d:\esmerald_venv\Lib\site-packages\lilya\routing.py", line 1582, in __call__
    await self.middleware_stack(scope, receive, send)
  File "d:\esmerald_venv\Lib\site-packages\lilya\routing.py", line 1422, in app
    await self.handle_route(route, path_handler=path_handler)
  File "d:\esmerald_venv\Lib\site-packages\lilya\routing.py", line 1325, in handle_route
    await route.handle_dispatch(path_handler.scope, path_handler.receive, path_handler.send)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "d:\esmerald_venv\Lib\site-packages\esmerald\routing\gateways.py", line 348, in handle_dispatch
    await self.app(scope, receive, send)
  File "d:\esmerald_venv\Lib\site-packages\esmerald\routing\router.py", line 2053, in __call__
    await self.handle_dispatch(scope=scope, receive=receive, send=send)
  File "d:\esmerald_venv\Lib\site-packages\esmerald\routing\router.py", line 2145, in handle_dispatch
    response = await self.get_response_for_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "d:\esmerald_venv\Lib\site-packages\esmerald\routing\base.py", line 509, in get_response_for_request
    response_data = await self._get_response_data(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "d:\esmerald_venv\Lib\site-packages\esmerald\routing\base.py", line 182, in _get_response_data
    data = await request_data
           ^^^^^^^^^^^^^^^^^^
  File "d:\esmerald_venv\Lib\site-packages\esmerald\transformers\model.py", line 177, in get_request_data
    form_data = await request.form()
                ^^^^^^^^^^^^^^^^^^^^
  File "d:\esmerald_venv\Lib\site-packages\lilya\requests.py", line 262, in _get_form
    self._form = await multipart_parser.parse()
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "d:\esmerald_venv\Lib\site-packages\lilya\_internal\_parsers.py", line 426, in parse
    parser.write(chunk)
  File "d:\esmerald_venv\Lib\site-packages\multipart\multipart.py", line 1101, in write
    l = self._internal_write(data, data_len)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "d:\esmerald_venv\Lib\site-packages\multipart\multipart.py", line 1207, in _internal_write
    raise e
multipart.exceptions.MultipartParseError: Did not find boundary character 49 at index 3

This is the traceback I encountered while attempting to upload a .csv file larger than 1MB.

**Reply by tarsil** [reply]
@prashikdewtale10 I understand but when you use the Form, the media type is not for multipart. What you need to do it's to replace the Form() with Body(media_type=EncodingType.MULTI_PART)) as per example above.

Can you do that and come back to me with the result, please?

This is not an Emerald limitation. Its HTTP protocol. 

Answering your question for best practices in general, you should have clear separate endpoints doing one thing and one thing only. 

Mixing a payload of plaintext and binary (file) can do more harm then good. 

Advice? Create different endpoints. One just for the file upload or uploads and another for data. That will also help you with testing.

**Reply by prashikdewtale10** [reply]
Hi @tarsil,

Thank you for your suggestion to replace Form() with Body(media_type=EncodingType.MULTI_PART). I implemented the changes as recommended, but unfortunately, I’m still encountering the same issue when uploading a .csv file larger than 1MB.

Here are the steps I followed:

Replaced Form() with Body(media_type=EncodingType.MULTI_PART) in my handler function.
Tested the API with a .csv file over 1MB using Postman.

**Reply by tarsil** [reply, marked as answer]
@prashikdewtale10 I literally added a test proving that passes with more than 1MB, maybe something its not right on your side? Let me share an example with you:

```python
import httpx

from io import BytesIO
from pathlib import Path
from typing import Any, Dict

from esmerald import Esmerald, Form, Gateway, Request, post

def create_dummy_file(size_in_bytes: int, filename: str = "test.txt"):
    file_buffer = BytesIO()
    file_buffer.write(b"0" * size_in_bytes)
    file_buffer.name = filename
    file_buffer.seek(0)
    return file_buffer


@post("/upload-data/")
async def create_upload_file(request: Request, data: Any = Form()) -> Dict[str, Any]:
        return {
            "filename": data.get("file").filename,
            "name": data.get("name"),
            "size": data.get("file").size,
        }


app = Esmerald(routes=[Gateway(handler=create_upload_file)])

# Now you can use httpx or requests to do something like this
data = {"name": "Test"}

seven_mb_file = create_dummy_file(size_in_bytes=7 * 1024 * 1024)
httpx.post('http://localhost:8000/upload-data',  files={"file": seven_mb_file}, data=data)
```

This works like a charm. A form that contains data and contains a file.
We call it "file" but since you are sending in a form, you can call whatever you want and need.

This is literally one of our tests.

**Reply by prashikdewtale10** [reply]
Hi @tarsil, thank you for your quick response, and I apologize for any confusion. I really appreciate the work you and the team have put into Esmerald. it’s an impressive framework! After further investigation, it seems the issue was with the Postman application I was using for API testing. When I executed the API using a Python script with urllib3 and curl, it worked properly without any issues.
Thank you for assistance and suggestions for best practices. 😃

**Reply by tarsil** [reply]
Not a problem my friend. I'm happy that you are using Esmerald. 

How do you feel so far with Esmerald? Production applications?
